source/fe/fe_p1nc.cc

```diff
@@ -0,0 +1,418 @@
+// ---------------------------------------------------------------------
+//
+// Copyright (C) 2015 - 2016 by the deal.II authors
+//
+// This file is part of the deal.II library.
+//
+// The deal.II library is free software; you can use it, redistribute
+// it, and/or modify it under the terms of the GNU Lesser General
+// Public License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE at
+// the top level of the deal.II distribution.
+//
+// ---------------------------------------------------------------------
+
+
+#include <deal.II/fe/fe_p1nc.h>
+
+DEAL_II_NAMESPACE_OPEN
+
+
+FE_P1NC::FE_P1NC()
+  :
+  FiniteElement<2,2>(FiniteElementData<2>(get_dpo_vector(),
+                                          1,
+                                          1,
+                                          FiniteElementData<2>::L2),
+                     std::vector<bool> (1, false),
+                     get_nonzero_component())
+{
+
+  // face support points: 2 end vertices
+  unit_face_support_points.resize(2);
+  unit_face_support_points[0][0] = 0.0 ;
+  unit_face_support_points[1][0] = 1.0 ;
+
+  // initialize constraints matrix
+  initialize_constraints () ;
+}
+
+
+
+std::string FE_P1NC::get_name () const
+{
+  return "FE_P1NC";
+}
+
+
+
+UpdateFlags FE_P1NC::requires_update_flags (const UpdateFlags flags) const
+{
+  UpdateFlags out = update_default;
+
+  if (flags & update_values)
+    out |= update_values | update_quadrature_points ;
+  if (flags & update_gradients)
+    out |= update_gradients ;
+  if (flags & update_cell_normal_vectors)
+    out |= update_cell_normal_vectors | update_JxW_values;
+  if (flags & update_hessians)
+    out |= update_hessians;
+
+  return out;
+}
+
+
+
+FiniteElement<2,2> *FE_P1NC::clone () const
+{
+  return new FE_P1NC(*this);
+}
+
+
+
+FE_P1NC::~FE_P1NC () {}
+
+
+
+std::vector<ComponentMask>
+FE_P1NC::get_nonzero_component()
+{
+  const unsigned int dofs_per_cell = 4;
+  std::vector<ComponentMask> masks (dofs_per_cell);
+  for (unsigned int i=0; i<dofs_per_cell; ++i)
+    masks[i] = ComponentMask(1,true);
+  return masks;
+}
+
+
+
+std::vector<unsigned int>
+FE_P1NC::get_dpo_vector ()
+{
+  std::vector<unsigned int> dpo(3);
+  dpo[0] = 1; // dofs per object: vertex
+  dpo[1] = 0; // line
+  dpo[2] = 0; // quad
+  return dpo;
+}
+
+
+
+void
+FE_P1NC::get_linear_shape (const Triangulation<2,2>::cell_iterator &cell,
+                           std::vector<double> &a,
+                           std::vector<double> &b,
+                           std::vector<double> &c)
+{
+  // edge midpoints
+  std::vector<Point<2> > mpt(4) ;
```

Comment:
since you know the size, it is more efficient (does not require dynamic memory allocation) if you just write it as
```
  Point<2> mpt[4];
```